### DISCUSSION TITLE
ArgumentValueResolver or ParamConverter?

### DISCUSSION THREAD

**Discussion by tjveldhuizen** [symfony/symfony#47312] [Q&A]
Given the [ambition](https://github.com/sensiolabs/SensioFrameworkExtraBundle/issues/776) to deprecate SensioFrameworkExtraBundle somewhere in the future, I am trying to find out what will be the replacement for ParamConverters. As far as I understand now, ArgumentValueResolvers are the way to go.

At a first glimpse, the conversion seems not to be very complicated: switch the interface on my existing ParamConverter, update the supports-method, replace the apply-method with resolve and tweak the implementations.

However: when `supports()` returns true, `resolve()` must [yield at least once](https://github.com/symfony/symfony/blob/56fc4b85481ec587543343eef6b877c04b788146/src/Symfony/Component/HttpKernel/Controller/ArgumentResolver.php#L65). This means that I have to make `supports()` return false if no object with the given slug exists in my database, thus I have to do the query twice: in `supports()` and in `resolve()`, or I have to cache the result in a property.

I'm afraid I don't understand the idea for this set-up, is it maybe not meant for the conversion of a slug to an entity?

**Comment by MatTheCat** [comment]
Sounds like you are talking about the new [`EntityValueResolver`](https://github.com/symfony/symfony/blob/6.2/src/Symfony/Bridge/Doctrine/ArgumentResolver/EntityValueResolver.php)? You can see it yields `null` or throw based on the argument.

**Reply by tjveldhuizen** [reply]
Thanks, I didn't know about it until somebody told me on Slack, yesterday. Thanks for the reference.

**Reply by MatTheCat** [reply]
You’re welcome. The gist is: if `supports` returns `true`, you **must** yield at least one value. If you cannot, then you **must** throw an exception.

**Reply by tjveldhuizen** [reply]
Plus: if the argument is optional, yielding `null` is OK/allowed

**Reply by MatTheCat** [reply]
See also https://github.com/symfony/symfony/pull/47363
